File: smtk/io/testing/cxx/ResourceSetTest.cxx
//=========================================================================
//  Copyright (c) Kitware, Inc.
//  All rights reserved.
//  See LICENSE.txt for details.
//
//  This software is distributed WITHOUT ANY WARRANTY; without even
//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//  PURPOSE.  See the above copyright notice for more information.
//=========================================================================

#include "smtk/resource/Set.h"

#include "smtk/attribute/Resource.h"

#include <iostream>

// Basic tests for smtk::resource::Set

int main(int /* argc */, const char* /* argv */ [])
{
  int status = 0;
  bool result;
  unsigned n;
  smtk::resource::Set resourceSet;

  // Create and add attribute resource
  smtk::attribute::ResourcePtr resource1 = smtk::attribute::Resource::create();
  result = resourceSet.add(resource1, "resource1", "", smtk::resource::Set::TEMPLATE);
  n = static_cast<unsigned>(resourceSet.numberOfResources());
  if (!result)
  {
    std::cerr << "add() call failed" << std::endl;
    status += 1;
  }
  else if (n != 1)
  {
    std::cerr << "Wrong number of resources: " << n << ", should be 1" << std::endl;
    status += 1;
  }

  // Create amd add 2nd attribute resource
  smtk::attribute::ResourcePtr resource2 = smtk::attribute::Resource::create();
  result = resourceSet.add(resource2, "resource2", "path2", smtk::resource::Set::INSTANCE);
  n = static_cast<unsigned>(resourceSet.numberOfResources());
  if (!result)
  {
    std::cerr << "add() call failed" << std::endl;
    status += 1;
  }
  else if (n != 2)
  {
    std::cerr << "Wrong number of resources: " << n << ", should be 2" << std::endl;
    status += 1;
  }

  // Add 1st resource w/different id and role
  result = resourceSet.add(resource1, "resource1-different-id", "", smtk::resource::Set::SCENARIO);
  n = static_cast<unsigned>(resourceSet.numberOfResources());
  if (!result)
  {
    std::cerr << "add() call failed" << std::endl;
    status += 1;
  }
  else if (n != 3)
  {
    std::cerr << "Wrong number of resources: " << n << ", should be 3" << std::endl;
    status += 1;
  }

  // Try using same id twice
  result = resourceSet.add(resource2, "resource2");
  n = static_cast<unsigned>(resourceSet.numberOfResources());
  if (result)
  {
    std::cerr << "add() call didn't fail" << std::endl;
    status += 1;
  }
  else if (n != 3)
  {
    std::cerr << "Wrong number of resources: " << n << ", should be 3" << std::endl;
    status += 1;
  }

  // Check resource ids
  std::vector<std::string> ids = resourceSet.resourceIds();
  if (ids.size() != 3)
  {
    std::cerr << "Wrong number of ids: " << ids.size() << ", should be 3" << std::endl;
    status += 1;
  }
  else
  {
    const char* expectedNames[] = { "resource1", "resource2", "resource1-different-id" };
    for (size_t i = 0; i < ids.size(); i++)
    {
      if (ids[i] != expectedNames[i])
      {
        std::cerr << "Wrong resource name " << ids[i] << ", should be " << expectedNames[i]
                  << std::endl;
        status += 1;
      }
    }
  }

  // Check resource info
  smtk::resource::Set::Role role;
  smtk::resource::Set::State state;
  std::string link;
  result = resourceSet.resourceInfo("resource2", role, state, link);
  if (!result)
  {
    std::cerr << "info() call failed" << std::endl;
    status += 1;
  }
  else
  {
    if (role != smtk::resource::Set::INSTANCE)
    {
      std::cerr << "Incorrect resource role " << role << ", should be "
                << smtk::resource::Set::INSTANCE << std::endl;
      status += 1;
    }
    if (state != smtk::resource::Set::LOADED)
    {
      std::cerr << "Incorrect resource state " << state << ", should be "
                << smtk::resource::Set::LOADED << std::endl;
      status += 1;
    }
    if (link != "path2")
    {
      std::cerr << "Incorrect resource link \"" << link << "\", should be \"path2\"" << std::endl;
      status += 1;
    }
  }

  // Retrieve resource
  smtk::resource::ResourcePtr resource;
  result = resourceSet.get("resource2", resource);
  if (!result)
  {
    std::cerr << "get() failed" << std::endl;
    status += 1;
  }

  std::cout << "Number of errors: " << status << std::endl;
  return status;
}

File: smtk/io/testing/python/ResourceSetTest.py
#!/usr/bin/python
import sys
#=============================================================================
#
#  Copyright (c) Kitware, Inc.
#  All rights reserved.
#  See LICENSE.txt for details.
#
#  This software is distributed WITHOUT ANY WARRANTY; without even
#  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
#  PURPOSE.  See the above copyright notice for more information.
#
#=============================================================================
from pprint import pprint
import smtk
if smtk.wrappingProtocol() == 'pybind11':
    import smtk.attribute
    import smtk.common
from smtk.simple import *

def RSTest():
    """Tests for ResourceSet
    Currently does not test for ResourceInfo
    as it is not implemented in the wrapper yet
    Otherwise this is a direct manual port of the cxx test"""

    status = 0
    result = False
    n = 0
    resourceSet = smtk.common.ResourceSet()

    system1 = smtk.attribute.System.New()
    print system1
    print system1.resourceType()
    result = resourceSet.addResource(system1, "system1", "", smtk.common.ResourceSet.TEMPLATE);

    n = resourceSet.numberOfResources()
    if result == False:
        print("addResource() call failed")
        status = status + 1
    elif n != 1:
        print("Wrong number of resources: %i, should be 1" % n)
        status = status + 1



    system2 = smtk.attribute.System.New()
    result = resourceSet.addResource(system2, "system2", "path2", smtk.common.ResourceSet.INSTANCE);

    n = resourceSet.numberOfResources()
    if result == False:
        print("addResource() call failed")
        status = status + 1
    elif n != 2:
        print("Wrong number of resources: %i, should be 2" % n)
        status = status + 1


    result = resourceSet.addResource(system1, "system1-different-id", "", smtk.common.ResourceSet.SCENARIO);
    n = resourceSet.numberOfResources()
    if result == False:
        print("addResource() call failed")
        status = status + 1
    elif n != 3:
        print("Wrong number of resources: %i, should be 3" % n)
        status = status + 1


    result = resourceSet.addResource(system2, "system2")
    n = resourceSet.numberOfResources()
    if result == True:
        print("addResource() call didn't fail failed")
        status = status + 1
    elif n != 3:
        print("Wrong number of resources: %i, should be 3" % n)
        status = status + 1


    ids = resourceSet.resourceIds()
    if len(ids) != 3:
        print("Wrong number of ids: %i, should be 3")
        status = status + 1
    else:
        expectedNames = ["system1", "system2", "system1-different-id"]
        for i in range(len(ids)):
            if ids[i] != expectedNames[i]:
                print("Wrong resource name %s, should be %s" % (ids[i], expectedNames[i]) )
                status = status + 1


    # Missing: ResourceInfo tests (function not implemented)

    # Note: ResourcePtr is not implemented (and cannot be due Resource being abstract -- shiboken issues)
    # Note: ResourceSet.get is modified by shiboken to return a ResourcePtr/shared_ptr<Resource>

    resource = resourceSet.get("system2")
    if resource == None:
        print("get() failed")
        status = status + 1
    rtype = resource.resourceType()
    if rtype != smtk.common.Resource.ATTRIBUTE:
        print("Incorrect resource type %s, should be smtk.common.Resource.ATTRIBUTE" % rtype)
        status = status + 1

    print("Number of errors: %i" % status)
    return status

def test():
    return RSTest()

if __name__ == '__main__':
    sys.exit(test())
